File: PieGallery/app/src/main/java/com/deadely/piegallery/view/favorites/presenter/FavoritesPresenter.java
package com.deadely.piegallery.view.favorites.presenter;

import com.deadely.piegallery.App;
import com.deadely.piegallery.base.BasePresenter;
import com.deadely.piegallery.database.AppDBHelper;
import com.deadely.piegallery.database.dao.FavoritesDao;
import com.deadely.piegallery.view.favorites.IFavoritesContract;

import javax.inject.Inject;

public class FavoritesPresenter extends BasePresenter<IFavoritesContract.View> implements IFavoritesContract.Presenter {
    private FavoritesDao favoritesDao;
    private AppDBHelper db;

    @Inject
    public FavoritesPresenter(FavoritesDao favoritesDao, AppDBHelper db) {
        this.favoritesDao = favoritesDao;
        this.db = db;
    }

    @Override
    public void showProgressBar() {
        getMvpView().showProgress();
    }

    @Override
    public void hideProgressBar() {
        getMvpView().hideProgress();
    }

    @Override
    public void clearList() {
        AppDBHelper db = App.getInstance().getDatabaseInstance();
        db.getFavoritesDao().deleteAll();

        getPhotoList();
    }

    @Override
    public void getPhotoList() {
        if (favoritesDao.getAllFavorites().size() == 0) {
            getMvpView().setEmptyList();
        } else {
            getMvpView().setPhotoList(favoritesDao.getAllFavorites());
        }
    }
}

File: PieGallery/app/src/main/java/com/deadely/piegallery/ui/favorites/presenter/FavoritesPresenter.kt
package com.deadely.piegallery.ui.favorites.presenter

import com.deadely.piegallery.base.BasePresenter
import com.deadely.piegallery.dataclasses.Photo
import com.deadely.piegallery.navigation.Screens
import com.deadely.piegallery.repository.Repository
import com.deadely.piegallery.ui.favorites.FavoritesView
import com.deadely.piegallery.utils.mapToModelList
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import moxy.InjectViewState
import javax.inject.Inject

@InjectViewState
class FavoritesPresenter @Inject constructor(private val repository: Repository) : BasePresenter<FavoritesView>() {
    fun exit() {
        router.replaceScreen(Screens.PhotosScreen())
    }

    fun getFavoritesPhotos() {
        viewState.showProgress()
        repository.getFavorites()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { data ->
                    viewState.setPhotos(data.mapToModelList())
                    viewState.hideProgress()
                },
                { e ->
                    viewState.showError()
                }
            )
    }

    fun deleteFromFavorites(photo: Photo) {
        repository.deleteFromFavorite(photo)
        getFavoritesPhotos()
    }

    fun openDetailScreen(photo: Photo) {
    }

    fun clearList() {
        repository.clearFavoriteList()
        getFavoritesPhotos()
    }
}
